### PR TITLE
Fix union type parameter inference

### LINKED ISSUE #5861: Generic of union type inferred as {}
(1.8.0-dev.20151201)

```typescript
interface Foo<T> { prop: T; }

declare function lift<U>(value: U | Foo<U>): Foo<U>;

function unlift<U>(value: U | Foo<U>): U {
	return lift(value).prop; // error TS2322: Type '{}' is not assignable to type 'U'.
}
```

In 1.6 and 1.7 `lift`'s generic is correctly inferred as `U`, so `prop` is of type `U` as expected. In 1.8 it's inferred as `{}`

Explicitly specifying the generic with `lift<U>(value).prop` works.

### PR DESCRIPTION
**Pull request by ahejlsberg**
Fixes #5861.

### CODE REVIEW

**Comment by ahejlsberg** [comment]
A bit more context on the fix. We previously would consider two type parameters *identical* if they were different type parameters but had identical constraints (but, oddly, wouldn't consider them assignable to each other). This of course isn't right, but it sort of worked when we were comparing constraints in generic signatures which was really the only time it would come up. However, we now sometimes need to compare type parameters for identity during union and intersection type inference (because of #5738) so we have to remove this logic. The effect is that we won't check whether type parameter constraints are identical when comparing generic signatures, as explained [here](https://github.com/Microsoft/TypeScript/pull/5895/files#diff-c3ed224e4daa84352f7f1abcd23e8ccaR5754).

File: tests/cases/compiler/unionTypeParameterInference.ts [hunk 1]
```diff
@@ -0,0 +1,7 @@
+interface Foo<T> { prop: T; }
+
+declare function lift<U>(value: U | Foo<U>): Foo<U>;
+
+function unlift<U>(value: U | Foo<U>): U {
+    return lift(value).prop; // error TS2322: Type '{}' is not assignable to type 'U'.
```
Comment (yuit) [inline tests/cases/compiler/unionTypeParameterInference.ts]: could you remove the error comment

**Comment by sandersn** [comment]
Could you also add a test for the case that type parameter constraints are *not* identical, but we don't check? It would be good to have as documentation in case do we go back and decide to start checking.

**Comment by sandersn** [comment]
:+1:

**Pull request merged**